hyperspy/_components/expression.py

```diff
@@ -0,0 +1,126 @@
+            the variable must be `x`. Also, in `module` is "numexpr" the
+            functions are limited to those that numexpr support. See its
+            documentation for details.
+        name : str
+            Name of the component.
+        position: str, optional
+            The parameter name that defines the position of the component if
+            applicable. It enables adjusting the position of the component
+            interactively in a model.
+        module: {"numpy", "numexpr"}, default "numpy"
+            Module used to evaluate the function. numexpr is often faster but
+            it supports less functions.
+
+        **kwarfs
+             Keyword arguments can be used to initialise the value of the
+             parameters.
+
+        Methods
+        -------
+        recompile: useful to recompile the function and gradient with a
+            a different module.
+
+        Examples
+        --------
+
+        The following creates a Gaussian component and set the initial value
+        of the parameters:
+
+        >>> hs.components.Expression(
+        ... expression="height * exp(-(x - x0) ** 2 * 4 * log(2)/ fwhm ** 2)",
+        ... name="Gaussian",
+        ... height=1,
+        ... fwhm=1,
+        ... x0=0,
+        ... position="x0",)
+
+        """
+
+        import sympy
+        self._str_expression = expression
+        self.recompile(module=module)
+        # Initialise component
+        Component.__init__(self, self._parameter_strings)
+        self.name = name
+        # Set the position parameter
+        if position:
+            self._position = eval("self.%s" % position)
+        # Set the initial value of the parameters
+        if kwargs:
+            for kwarg, value in kwargs.iteritems():
+                exec("self.%s.value = value" % kwarg)
+
+        self.__doc__ = _CLASS_DOC % (name,
+                                     sympy.latex(sympy.sympify(expression)))
+
+    def function(self, x):
+        return eval(self._f_eval_str)
+
+    def recompile(self, module="numpy"):
+        import compiler
+        import sympy
+        from sympy.utilities.lambdify import lambdify
+        expr = sympy.sympify(self._str_expression)
+        eval_expr = expr.evalf()
+        # Extract parameters
+        parameters = [
+            symbol for symbol in expr.free_symbols if symbol.name != "x"]
+        # Extract x
+        x, = [symbol for symbol in expr.free_symbols if symbol.name == "x"]
+        # Create compiled function
+        self._f = lambdify([x] + parameters, eval_expr,
+                           modules=module, dummify=False)
+        parnames = [symbol.name for symbol in parameters]
+        eval_str = "self._f(x, %s)" % ", ".join(
+            ["self.%s.value" % par for par in parnames])
+        # Generate string to be evaluated by self.function
+        self._f_eval_str = compiler.compile(eval_str, '<string>', 'eval')
```

Comment:
Two good points, thanks! I've fixed the compiler bit.

Regarding `eval`, I  agree that it can be dangerous. I would be worried if we were saving the strings that we pass to `eval` in the hdf5 file, in which case it would indeed be possible to do something nasty just by loading the file using HyperSpy. However, only the expression in string format as passed to the constructor will be stored in the hdf5 file, so one may think that there is not much room for evilness here (at least there is no increased risk over running any other python script). However, one will be very wrong! `sympy.sympify` uses `eval` internally (!) I don't see any way out. We could require `expression` to be SymPy expression instead of a string, but we cannot save SymPy expressions in hdf5, so we have to turn it into a string at some point... An inconvenient workaround could be to print the string before executing sympyfy and ask the user to confirm the action. What do you think?